Fixes for PHP 7.2 compatibility

SilverFire opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | no, but will
| Fixed issues  | #14489

SilverFire committed:
Fixes for PHP 7.2 compatibility

SilverFire commented:
There are still issues with `session_cache_limiter()`
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/filters/HttpCache.php#L192

And with `session_set_cookie_params()`
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/web/Session.php#L381

They work correctly on PHP 7.1, but not on 7.2. Is seems that PHPUnit does not restore environment after tests correctly. Maybe it is fixed in PHPUnit 6.*, but it requires PHP 7 or higher. 

I would not like to skip these tests on PHP 7.2, but I don't have a clue how to fix it. I will be happy to hear some suggestions @yiisoft/reviewers @yiisoft/core-developers

SilverFire commented:
Checked with PHPUnit 6.4.1 on PHP 7.2 – the problem is still present.

rob006 commented:
There is a issue about `session_cache_limiter()`: https://github.com/yiisoft/yii2/issues/14811

SilverFire commented:
Assuming our rare releases schedule, I'd prefer to put 7.2 compatibility fixes in 2.0.13

SilverFire commented:
> When headers are already sent and try to set new INI values, session_name(),
      session_module_name(), session_save_path(), session_cache_limiter() and
      session_cache_expire() are no longer works. Older PHPs accepts new values even
      if new values will not be effective.
      This new corrected behavior may affect command line mode CLI scripts that manage
      sessions. Use output buffer just like web applications to resolve problems on
      CLI scripts.

From PHP 7.2 [upgrade notes](https://github.com/php/php-src/blob/php-7.2.0RC4/UPGRADING)

bscheshirwork commented:
> This new corrected behavior may affect command line mode CLI scripts that manage
sessions. Use output buffer just like web applications to resolve problems on
CLI scripts.

This is only sessions, not headers?
We need change it?
https://github.com/yiisoft/yii2/blob/277468a7aa7c182ca7a686d3d82a2e38d897a980/framework/base/ErrorHandler.php#L100-L104

SilverFire committed:
More fixes for PHP 7.1 compatibility

SilverFire committed:
Added ResultPrinter to send output to STDOUT insted of printing it

Method `PHPUnit\Util\Printer::write($buffer)` on line 112 prints buffer and causes errors related to headers
that can not be sent when output is already started

SilverFire commented:
Partially fixed, but there is a mess with date parsing. Don't have enough time to check it now. If anybody can - it will be good

klimov-paul commented on tests/framework/i18n/FormatterTest.php:
```diff
@@ -293,18 +294,26 @@ public function lengthDataProvider()
      * @dataProvider lengthDataProvider
      */
-    public function testIntlAsLength($message, $arguments, $expected)
+    public function testIntlAsLength($message, $arguments, $expected, $_, $expectedException = [])
```
Why variable named `$_`?

SilverFire replied:
It contains shotLength expectation and is unused in this test.

klimov-paul replied:
Fact that it is unused does not justifies such variable naming.

klimov-paul commented on framework/filters/HttpCache.php:
```diff
@@ -182,6 +182,10 @@ protected function validateCache($lastModified, $etag)
      */
     protected function sendCacheControlHeader()
     {
+        if (headers_sent($file, $line)) {
```
Is this check really necessary? Would not PHP throw its own error saying 'headers have already been sent'?

klimov-paul commented on tests/ResultPrinter.php:
```diff
@@ -0,0 +1,29 @@
+/**
+ * IsOneOfAssert asserts that the value is one of the expected values.
+ */
+class ResultPrinter extends \PHPUnit\TextUI\ResultPrinter
```
Using custom `ResultPrinter` as a solution, does not seem right.
What if some developer creates component or extension, which operates session and wishes to write a unit tests for it? He will have to create his own `ResultPrinter` as well, which does not feel right.

We should ensure framework testability.

bscheshirwork replied:
+1
I must been redefine PHPStorm printer - his realization send headers too.
```php
class PhpStorm_Codeception_ReportPrinter_Redefine extends PHPUnit_TextUI_ResultPrinter
{
    /**
     * @inheritDoc
     */
    public function __construct(
        $out = null,
        $verbose = false,
        $colors = self::COLOR_DEFAULT,
        $debug = false,
        $numberOfColumns = 80,
        $reverse = false
    ) {
        parent::__construct(STDOUT, $verbose, $colors, $debug, $numberOfColumns, $reverse);
    }

    /**
     * Flush buffer and close output if it's not to a STDOUT stream
     */
    public function flush()
    {
        if ($this->out != STDOUT) {
            parent::flush();
        }
    }
}

class PhpStorm_Codeception_ReportPrinter extends PhpStorm_Codeception_ReportPrinter_Redefine
...
```

SilverFire replied:
It's not a final solution, I'm just trying to make it work somehow :)
There is an issue in PHPUnit repo https://github.com/sebastianbergmann/phpunit/issues/1193

bscheshirwork replied:
@SilverFire yes, I tell about this problem. p.s.: Don't forgot `flush()`

SilverFire replied:
Done

SilverFire committed:
Disabled INTL tests in PHP 7.2

SilverFire committed:
Use INTL_IDNA_VARIANT_UTS46 for idn_to_ascii because of deprecations

https://github.com/php/php-src/pull/2355

SilverFire committed:
Fixed HtmlTest::testCsrfMetaTagsEnableCsrfValidationWithoutCookieValidationKey()

SilverFire committed:
Fixed count call on non-array in \yii\web\User::getIdentityAndDurationFromCookie()

SilverFire committed:
Disabled XML mime-type test in FileValidatorTest because of PHP regression

https://bugs.php.net/bug.php?id=75380

SilverFire commented:
Reported 2 regressions in PHP core:

https://bugs.php.net/bug.php?id=75378
https://bugs.php.net/bug.php?id=75380

All tests except DB are passing now.

SilverFire committed:
Removed excess check, introduces during PHP 7.2 compatibility fixes session

SilverFire committed:
minor

SilverFire committed:
minor

samdark commented on tests/ResultPrinter.php:
```diff
@@ -0,0 +1,29 @@
+namespace yiiunit;
+
+/**
+ * IsOneOfAssert asserts that the value is one of the expected values.
```
phpdoc looks wrong.

samdark replied:
phpunit issue should be referenced. It's not clear about the purpose of the printer from the code itself.

SilverFire replied:
Actually, it's not a PHPUnit issue.
PHP have changed `session_cache_limiter()` so it does not work when headers are already sent.
https://github.com/php/php-src/blob/php-7.2.0RC4/UPGRADING#L89

So our tests are broken. Maybe it will be better to mock `session_cache_limiter()` call instead of changing printing output

samdark commented on framework/web/Session.php:
```diff
@@ -376,6 +376,9 @@ public function setCookieParams(array $value)
      */
     private function setCookieParamsInternal()
     {
+        if (headers_sent($file, $line)) {
```
Won't `session_set_cookie_params()` trigger the same exception natively?

samdark commented on tests/framework/helpers/HtmlTest.php:
```diff
@@ -140,12 +140,17 @@ public function testCsrfMetaTagsEnableCsrfValidation()
+        ]);
+        $this->expectException('yii\base\InvalidConfigException');
+        $this->expectExceptionMessage('yii\web\Request::cookieValidationKey must be configured with a secret key.');
+        Html::csrfMetaTags();
```
Isn't assertion with a pattern missing here?

SilverFire replied:
No pattern will be generated - exception will be thrown instead.

cebe replied:
I see now. The mock generates a class that does not work as it normally should.

samdark commented on tests/framework/i18n/IntlTestHelper.php:
```diff
@@ -24,10 +24,18 @@ public static function setIntlStatus($test)
             if (strncmp($test->getName(false), 'testIntl', 8) === 0) {
+                static::$enableIntl = true;
+
+                if (strpos(PHP_VERSION, '7.2.0RC') !== false && version_compare(PHP_VERSION, '7.2.0.RC.3', '<=')) {
```
Why using both strpos and version_compare?

samdark commented on tests/framework/validators/FileValidatorTest.php:
```diff
@@ -388,15 +388,19 @@ public function testValidateMimeTypeMaskInvalid($fileName, $mask)
-            ['test.xml', '*/xml', 'xml'],
+            // Disabled for PHP 7.2 RC because of regression:
+            // https://bugs.php.net/bug.php?id=75380
+            version_compare(PHP_VERSION, '7.2.0', '>=') && strpos(PHP_VERSION, 'RC')
```
Wouldn't it skip all RCs for 7.2, 7.3 etc?

SilverFire committed:
Fixed PHPUnit compatibility

SilverFire committed:
Removed echo statement to prevent error "headers have already been sent"

SilverFire committed:
More PHPUnit compatibility

SilverFire committed:
Enhanced PHP version testing conditions, other minor changes;

SilverFire committed:
Removed PHPUnit <6.0 compatibility for some tests of ArrayHelper

SilverFire committed:
Fixed Email and URL validators compatibility with older PHP versions

SilverFire committed:
One more try to fix tests

SilverFire committed:
Started DB layer adjustments to support PHP 7.2

SilverFire committed:
Trying to fix tests on Travis

SilverFire committed:
trying to fix moar PHP 7.2 errors

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Anoter try to fix tests

SilverFire committed:
Comment to travis config

SilverFire commented:
Fixed all remaining tests. Ready for review and merge. @yiisoft/reviewers

cebe commented on .travis.yml:
```diff
@@ -149,6 +148,7 @@ before_script:
 
   # show some versions and env information
   - php --version
+  - php --ini
```
looks like this breaks HHVM tests, it starts the binary which is waiting for input.

SilverFire committed:
Fixed HHVM tests, enhanced ActiveRecordTestTrait

SilverFire commented:
Thanks, Carsten. Fixed.

> how and why was this working before?

I was surprised as well.

samdark commented on framework/validators/UrlValidator.php:
```diff
@@ -108,6 +108,16 @@ protected function validateValue($value)
+    private function idnToAscii($idn)
+    {
+        if (PHP_VERSION_ID < 50600) {
+            // TODO: drop old PHP versions support
```
2.1?

SilverFire replied:
Yes

rob006 committed:
Fix PHPDoc

rob006 commented on tests/framework/helpers/ArrayHelperTest.php:
```diff
@@ -810,20 +810,22 @@ public function testGetValueObjects()
      */
     public function testGetValueNonexistingProperties1()
     {
+        $this->expectException('PHPUnit\Framework\Error\Notice');
```
We could use `class_exists()` to detect expected exception and make this test work on phpunit <6.0

SilverFire replied:
I think it's not so important and could be skipped

samdark commented:
That was big! Thank you for patience. Now we're 7.2 ready! Yay!

bscheshirwork commented:
We have fix in `php-7.2.0RC5`
[75378](https://github.com/php/php-src/blob/fcbf63d8087db46f2ecd9248687ef2a463013a14/NEWS#L31-L33)

SilverFire commented:
Already encountered in https://github.com/yiisoft/yii2/commit/336404e0b992406732d80bb239ecf3780dda4359
